src/webgpu/api/operation/command_buffer/render/index_format.spec.ts

```diff
@@ -0,0 +1,299 @@
+  private result!: GPUBuffer;
+  private swapChainFormat: GPUTextureFormat = 'bgra8unorm';
+
+  async initResources(format: GPUIndexFormat): Promise<void> {
+    // glslang module
+    this.glslang = await glslangModule();
+
+    // render pipeline
+    this.pipeline = this.MakeRenderPipeline(format);
+
+    const context = this.CreateRenderContext(true);
+    if (context !== null) {
+      const swapChain: GPUSwapChain = context.configureSwapChain({
+        device: this.device,
+        format: this.swapChainFormat,
+        usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.OUTPUT_ATTACHMENT,
+      });
+      this.colorAttachment = swapChain.getCurrentTexture();
+    }
+
+    // result buffer
+    this.result = this.device.createBuffer({
+      size: 4,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+  }
+
+  CreateRenderContext(display: boolean): any {
+    const canvas = document.createElement('canvas');
+    canvas.width = 100;
+    canvas.height = 100;
+    // Display rendering on test page
+    if (display) {
+      const container = document.getElementById('canvasContainer');
+      if (container) {
+        container.innerHTML = '';
+        container.append(canvas);
+      }
+    }
+    return canvas.getContext('gpupresent');
+  }
+
+  MakeRenderPipeline(format: GPUIndexFormat): GPURenderPipeline {
+    const vertexShaderGLSL = `#version 450
+    layout(location = 0) in vec4 pos;
+    void main() {
+        gl_Position = pos;
+    }`;
+
+    const fragmentShaderGLSL = `#version 450
+    layout(location = 0) out vec4 fragColor;
+    void main() {
+        fragColor = vec4(0.0, 1.0, 0.0, 1.0);
+    }`;
+
+    return this.device.createRenderPipeline({
+      layout: this.device.createPipelineLayout({ bindGroupLayouts: [] }),
+
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: this.glslang.compileGLSL(vertexShaderGLSL, 'vertex'),
+        }),
+        entryPoint: 'main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: this.glslang.compileGLSL(fragmentShaderGLSL, 'fragment'),
+        }),
+        entryPoint: 'main',
+      },
+
+      primitiveTopology: 'triangle-strip',
+
+      colorStates: [
+        {
+          format: this.swapChainFormat,
+        },
+      ],
+
+      vertexState: {
+        indexFormat: format,
+        vertexBuffers: [
+          {
+            arrayStride: 4 * 4,
+            stepMode: 'vertex',
+            attributes: [
+              {
+                format: 'float4',
+                offset: 0,
+                shaderLocation: 0,
+              },
+            ],
+          },
+        ],
+      },
+    });
+  }
+
+  MakeBufferMapped(arrayBuffer: ArrayBuffer, usage: number): GPUBuffer {
+    const [buffer, bufferMapping] = this.device.createBufferMapped({
+      size: arrayBuffer.byteLength,
+      usage,
+    });
+
+    if (arrayBuffer instanceof Float32Array) {
```

Comment:
Won't these always be false?

Reply:
I tested it's true if arrayBuffer = new Float32Array()

Reply:
Actually, it's just the type annotation of `arrayBuffer` looks wrong.

I think you should remove this helper and inline the code where you use it now; the inline sites should be quite small.